week7/src/components/CommonPage.tsx

```diff
@@ -0,0 +1,113 @@
+import styled from "styled-components";
+import { useEffect, useMemo, useState } from "react";
+import SingleCard from "./SingleCard";
+import Header from "./Header";
+import Modal from "./Modal";
+import ModalPortal from "./ModalPortal";
+import { useRecoilState } from "recoil";
+import { scoreState, modalOnState, disabledState } from "../atom/atom";
+
+export interface CardImgListProps {
+  name?: string;
+  alt?: string;
+  matched: boolean;
+}
+
+export interface CommonPageProps {
+  cardList: CardImgListProps[];
+  shuffle: number;
+}
+
+const CommonPage = (props: CommonPageProps) => {
+  const { cardList, shuffle } = props;
+
+  const [choiceOne, setChoiceOne] = useState<CardImgListProps | null>(null);
+  const [choiceTwo, setChoiceTwo] = useState<CardImgListProps | null>(null);
+  const [score, setScore] = useRecoilState(scoreState);
+  const [modalOn, setModalOn] = useRecoilState(modalOnState);
+  const [disabled, setDisabled] = useRecoilState(disabledState);
+  const flippedCard = document.getElementsByClassName("flipped");
+
+  const copiedCardList = useMemo(() => {
+    const copied = JSON.parse(
+      JSON.stringify(
+        (Object.keys(cardList) as Array<keyof typeof cardList>)
+          .map((item) => cardList[item])
+          .flat()
+      )
+    );
+    copied.map((it: CardImgListProps) => (it.matched = false));
+
+    return copied.sort(() => Math.random() - 0.5);
+  }, [cardList, shuffle]);
+
+  const handleChoice = (card: CardImgListProps) => {
+    choiceOne ? setChoiceTwo(card) : setChoiceOne(card);
+  };
+
+  useEffect(() => {
+    setScore(0);
+    resetCardSet();
+  }, [copiedCardList]);
+
+  useEffect(() => {
+    if (choiceOne && choiceTwo) {
+      setDisabled(true);
+
+      if (choiceOne.name === choiceTwo.name) {
+        choiceOne.matched = true;
+        choiceTwo.matched = true;
```

Comment:
새로운 값을 직접 할당하려고 해서 저 readOnly 에러가 발생한 것 같은데!!
이렇게 새 객체를 만들고 업데이트 된 값을 할당하면 직접 수정하지 않고 안전하게 업데이트 할 수 있어!!

```suggestion
  const updatedChoiceOne = { ...choiceOne, matched: true };
  const updatedChoiceTwo = { ...choiceTwo, matched: true };

  setChoiceOne(updatedChoiceOne);
  setChoiceTwo(updatedChoiceTwo);
```